### PR TITLE
[fix](multi-table-load) fix memory leak when processing multi-table routine load

### PR DESCRIPTION
**Pull request by Hastyshell**
## Proposed changes

close:#21606
 
### Debug

Using jemalloc prof tool and found there is unreasonable memory overhead within multi_table_put_result obj.
image

Found that there are cycle refs during processing each single routine load task, i.e. StreamLoadContext holds a shared ptr ref to MultiTablePipe, and MultitablePipe holds shared ptr ref to StreamLoadContext as well.

Furthermore, during related loading process, StreamLoadContext's obj cnt is always growing and do not decrease when load stream stopped.
image

### Resolve

Using a naked ptr rather than shared ptr in MultiTablePipe to hold the StreamLoadContext, so that both the StreamLoadContext and the MultiTablePipe will be released normally. 

Why not use weak_ptr?
1. Cuz introducing weak_ptr is not so elegant and has lower performance.
2. MultiTablePipe's life cycle is under control of the StreamLoadContext, it is reasonable to use raw ptr to provide MultiTablePipe with an access to the current ctx. StreanLoadContext is sure to be released after MultiTablePipe, and will not result in dangling pointer. 

### Fix result

Memory usage during load process:
image

Observing the obj cnt:
image

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by TangSiyang2001** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 57.09 seconds
 stream load tsv:          501 seconds loaded 74807831229 Bytes, about 142 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          89.3 seconds inserted 10000000 Rows, about 111K ops/s
 storage size: 17164185513 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230707130115_clickbench_pr_174217.html

**Review by freemandealer — APPROVED** [review]
LGTM

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 52.29 seconds
 stream load tsv:          447 seconds loaded 74807831229 Bytes, about 159 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          57 seconds loaded 1101869774 Bytes, about 18 MB/s
 stream load parquet:          29 seconds loaded 861443392 Bytes, about 28 MB/s
 insert into select:          69.9 seconds inserted 10000000 Rows, about 143K ops/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230707060856_clickbench_pr_174231.html

File: be/src/io/fs/multi_table_pipe.h [hunk 1]
```diff
@@ -74,7 +74,7 @@ class MultiTablePipe : public KafkaConsumerPipe {
     std::atomic<uint64_t> _unplanned_row_cnt {0}; // trigger plan request when exceed threshold
     std::atomic<uint64_t> _inflight_plan_cnt {0}; // how many plan fragment are executing?
     std::atomic<bool> _consume_finished {false};
-    std::shared_ptr<StreamLoadContext> _ctx;
+    StreamLoadContext* _ctx;
```
Comment (airborne12) [inline be/src/io/fs/multi_table_pipe.h:77]: No safe to use row pointer
Is there a cycle reference for StreamLoadContext shared_ptr which causes memory is not deallocated?

Comment (TangSiyang2001) [reply be/src/io/fs/multi_table_pipe.h:77]: @airborne12 
Yes it is, for not using weak_ptr to solve the problem:

> Why not use weak_ptr?

> Cuz introducing weak_ptr is not so elegant and has lower performance.
> MultiTablePipe's life cycle is under control of the StreamLoadContext, it is reasonable to use naked ptr to 
> provide MultiTablePipe with an access to the current ctx.

**Review by dataroaring — DISMISSED** [review]
LGTM

**Comment by freemandealer** [comment]
run p1

File: be/src/io/fs/multi_table_pipe.h (hunk 1 above)
Comment (yiguolei) [inline be/src/io/fs/multi_table_pipe.h:77]: Please add comment here to explain why using raw pointer here. Or other people may refactor code and using shared ptr again.

Comment (TangSiyang2001) [reply be/src/io/fs/multi_table_pipe.h:80]: modified

**Comment by TangSiyang2001** [comment]
run buildall

**Review by yiguolei — APPROVED** [review]
LGTM

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 56.88 seconds
 stream load tsv:          503 seconds loaded 74807831229 Bytes, about 141 MB/s
 stream load json:         19 seconds loaded 2358488459 Bytes, about 118 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          86.7 seconds inserted 10000000 Rows, about 115K ops/s
 storage size: 17162098002 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230711235148_clickbench_pr_176661.html

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 52.25 seconds
 stream load tsv:          447 seconds loaded 74807831229 Bytes, about 159 MB/s
 stream load json:         21 seconds loaded 2358488459 Bytes, about 107 MB/s
 stream load orc:          57 seconds loaded 1101869774 Bytes, about 18 MB/s
 stream load parquet:          28 seconds loaded 861443392 Bytes, about 29 MB/s
 insert into select:          70.0 seconds inserted 10000000 Rows, about 142K ops/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230711165712_clickbench_pr_176662.html

**Comment by TangSiyang2001** [comment]
run p0

**Review by kaijchen — APPROVED** [review]
LGTM

**Pull request merged**